### DISCUSSION TITLE
What are the RF modules supported by MicroPython for long range audio communication?

### DISCUSSION THREAD

**Discussion by th4ruka** [micropython/micropython#14464] [General]
Hi!
I'm planning to build a pair of audio transceiver devices using 2 `ESP32`s and 2 RF wireless modules. I want your help to decide which RF modules to use. I have some doubts about selecting the best RF module for the project.

So, I'm really grateful if someone can help me with these problems,
1. What are the best suited RF modules for long range communication about 1 km - 2 km range?
2. Can I use built-in WiFi module of `ESP32` with `ESP-NOW` protocol?
3. Is it required to have a specific MicroPython library for a RF module or can I use existing protocols like `I2C`, `SPI` to communicate with the module?

**Comment by davefes** [comment]
For a simplex PTT (push-to-talk) link I would start searching for info using LoRa.  I have seen articles about using ESP32 and various radio chips but have not investigated them.

1) LoRa?
2) Yes
3) Yes, unless you are an experienced driver writer.  Setting up a LoRa chip is a BIG task.

**Reply by peterhinch** [reply]
To add to the above, ESP-NOW works very well but I don't think you'll get to 1km, at least not without highly directional antennas.

**Reply by th4ruka** [reply]
Yes, I think the main drawback of ESP-NOW is its limited range, which is about 500 meters (line of sight). Trying directional antennas could be a good idea to extend the range.
As I know, LoRa can achieve a 1 km range, but its drawback is the low bandwidth, resulting low audio quality.

**Reply by Josverl** [reply]
Assumption : for a PTT audio stream you will need approx 18 Kbps 

Lora can provide about 5 Kbps , so with 4 transmitters to 4 receivers and good antennas you may be able to get what you need 
Multiplexing / demultiplexing will be a chore/challenge, but DMA may be able to help there.

there are some industrial solutions revealed by a quick search, that show that it can be done :

- https://www.satel.com/products/radio-modems/satel-xprs-ip-radio-router/
- https://oleumtech.com/oleumtech-products/io-max-transmitter-mx1

**Reply by davefes** [reply]
https://unsigned.io/understanding-lora-parameters/
BW 500 kHz, SF 7, CR 4:5 gives 21.88kbps

Maybe, check-out:
https://www.youtube.com/watch?v=rj9GZQtFs8k&t=889s
WiFi HaLow

**Comment by victorallume** [comment]
You could consider normal wifi with a cantenna or other highly-directional antenna. People have gotten 10s of kms with such a setup (though that will also depend on RF transmit power of the module). You do need line-of-sight, and a way to align the antennas.
